### PR TITLE
Fix light client

### PR DESCRIPTION
**Pull request by junha1**
Previous algorithm has a critical problem.

### CODE REVIEW

File: rpc/src/v1/impls/ibc.rs [hunk 1]
```diff
@@ -102,28 +102,33 @@ where
...
-        let state = self.client.state_at(block_id).unwrap();
-
+        if self.client.state_at(block_id_next).is_none() {
+            return Ok(None)
+        }
+        let state_prev = self.client.state_at(block_id_prev).unwrap();
         let header_core = self.client.block_header(&block_id).unwrap();
-        let vset_raw = ccore::stake::NextValidators::load_from_state(&state).map_err(errors::core)?;
+        let header_core_next = self.client.block_header(&block_id_next).unwrap();
+        let vset_raw = ccore::stake::NextValidators::load_from_state(&state_prev).map_err(errors::core)?;
```
Comment (majecty) [inline rpc/src/v1/impls/ibc.rs]: If we use `CurrentValidators`, we don't need to read the previous block.

File: rpc/src/v1/impls/ibc.rs [hunk 2]
```diff
@@ -102,28 +102,33 @@ where
     }
 
     fn compose_header(&self, block_number: u64) -> Result<Option<Bytes>> {
+        // FIXME: Currently compose_header() works only for at most bestblock-1
```
Comment (majecty) [inline rpc/src/v1/impls/ibc.rs]: Could you give more context?
Like "block n's seal is stored in the next block's header".

File: rpc/src/v1/impls/ibc.rs (hunk 1 above)
Comment (junha1) [reply rpc/src/v1/impls/ibc.rs]: Ok, then I'll add `create_compact_validator_set()` on `CurrentValidators` and use that.

File: rpc/src/v1/impls/ibc.rs (hunk 2 above)
Comment (junha1) [reply rpc/src/v1/impls/ibc.rs]: Ok

**Comment by junha1** [comment]
@majecty Updated.

File: core/src/consensus/light_client/mod.rs [hunk 3]
```diff
@@ -44,10 +44,11 @@ impl ClientState {
 /// All data for updating a light client up to Nth block.
```
Comment (majecty) [inline core/src/consensus/light_client/mod.rs]: For the " Fix invalid light client core algorithm " commit, please write these things in your commit message:
* What the problem was: number was not verified.
* How you fixed the problem: include a raw header in the "UpdateHeader".

File: core/src/ibc/client_02/client.rs [hunk 4]
```diff
@@ -14,6 +14,7 @@
 // You should have received a copy of the GNU Affero General Public License
```
Comment (majecty) [inline core/src/ibc/client_02/client.rs]: For the commit message: " Adapt ICS02 client to the new core light client algorithm", I don't understand the meaning of the message. Could you explain it to me?

File: rpc/src/v1/impls/ibc.rs [hunk 5]
```diff
@@ -102,21 +102,29 @@ where
     }
```
Comment (majecty) [inline rpc/src/v1/impls/ibc.rs]: For the commit message "Fix to provide correct seal and validator set", please write what was the problem and how you fixed it.

File: core/src/ibc/client_02/client.rs (hunk 4 above)
Comment (junha1) [reply core/src/ibc/client_02/client.rs]: @majecty 
Means ICS 02 has to be changed along the change of core light client algorithm.

Comment (majecty) [reply core/src/ibc/client_02/client.rs]: Did you mean that we should change our ICS02 light client implementation along to the change of the light client you changed at the previous commit?
If it is right, please merge it with the light client commit or give a brief description about what was changed in the light client code.

Comment (junha1) [reply core/src/ibc/client_02/client.rs]: I think it's better to just squash

**Comment by junha1** [comment]
@majecty  Updated

**Pull request merged**